File: e2e/go/links_test.go
// E2e tests for category: links
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_LinksAnchorFragment(t *testing.T) {
	// Identifies fragment-only links as anchor type
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if !strings.Contains(result.Links[""].LinkType, `anchor`) {
		t.Errorf("expected to contain %s", `anchor`)
	}
}

func Test_LinksBaseTag(t *testing.T) {
	// Resolves relative URLs using base tag href
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Links) <= 2 {
		t.Errorf("expected > 2, got %v", len(result.Links))
	}
	if !strings.Contains(result.Links[""].Url, `example.com`) {
		t.Errorf("expected to contain %s, got %q", `example.com`, result.Links[""].Url)
	}
}

func Test_LinksDocumentTypes(t *testing.T) {
	// Detects PDF, DOCX, XLSX links as document type
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if !strings.Contains(result.Links[""].LinkType, `document`) {
		t.Errorf("expected to contain %s", `document`)
	}
}

func Test_LinksEmptyHref(t *testing.T) {
	// Handles empty href attributes without errors
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Links) <= 0 {
		t.Errorf("expected > 0, got %v", len(result.Links))
	}
	if !strings.Contains(result.Links[""].Url, `/valid`) {
		t.Errorf("expected to contain %s, got %q", `/valid`, result.Links[""].Url)
	}
}

func Test_LinksInternalExternalClassification(t *testing.T) {
	// Correctly classifies internal vs external links by domain
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Links) <= 4 {
		t.Errorf("expected > 4, got %v", len(result.Links))
	}
	if !strings.Contains(result.Links[""].LinkType, `internal`) {
		t.Errorf("expected to contain %s", `internal`)
	}
	if !strings.Contains(result.Links[""].LinkType, `external`) {
		t.Errorf("expected to contain %s", `external`)
	}
}

func Test_LinksMailtoJavascriptSkip(t *testing.T) {
	// Skips mailto:, javascript:, and tel: scheme links
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Links) <= 0 {
		t.Errorf("expected > 0, got %v", len(result.Links))
	}
	if strings.Contains(result.Links[""].Url, `mailto:`) {
		t.Errorf("expected NOT to contain %s, got %q", `mailto:`, result.Links[""].Url)
	}
}

func Test_LinksProtocolRelative(t *testing.T) {
	// Handles protocol-relative URLs (//example.com) correctly
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Links) <= 1 {
		t.Errorf("expected > 1, got %v", len(result.Links))
	}
	if len(result.Links[""].ProtocolRelative) == 0 {
		t.Errorf("expected non-empty value")
	}
}

func Test_LinksRelAttributes(t *testing.T) {
	// Preserves rel=nofollow and rel=canonical attributes
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Links) <= 0 {
		t.Errorf("expected > 0, got %v", len(result.Links))
	}
}

func Test_LinksRelativeParent(t *testing.T) {
	// Resolves ../ and ./ relative parent path links correctly
	result, err := pkg.Scrape()
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Links) <= 3 {
		t.Errorf("expected > 3, got %v", len(result.Links))
	}
}

File: e2e/elixir/test/links_test.exs
# E2e tests for category: links
defmodule E2e.LinksTest do
  use ExUnit.Case, async: true

  describe "links_anchor_fragment" do
    test "Identifies fragment-only links as anchor type" do
      result = Kreuzcrawl.scrape!()
      assert String.contains?(result.links.get("").link_type, "anchor")
    end
  end

  describe "links_base_tag" do
    test "Resolves relative URLs using base tag href" do
      result = Kreuzcrawl.scrape!()
      assert length(result.links) > 2
      assert String.contains?(result.links.get("").url, "example.com")
    end
  end

  describe "links_document_types" do
    test "Detects PDF, DOCX, XLSX links as document type" do
      result = Kreuzcrawl.scrape!()
      assert String.contains?(result.links.get("").link_type, "document")
    end
  end

  describe "links_empty_href" do
    test "Handles empty href attributes without errors" do
      result = Kreuzcrawl.scrape!()
      assert length(result.links) > 0
      assert String.contains?(result.links.get("").url, "/valid")
    end
  end

  describe "links_internal_external_classification" do
    test "Correctly classifies internal vs external links by domain" do
      result = Kreuzcrawl.scrape!()
      assert length(result.links) > 4
      assert String.contains?(result.links.get("").link_type, "internal")
      assert String.contains?(result.links.get("").link_type, "external")
    end
  end

  describe "links_mailto_javascript_skip" do
    test "Skips mailto:, javascript:, and tel: scheme links" do
      result = Kreuzcrawl.scrape!()
      assert length(result.links) > 0
      refute String.contains?(result.links.get("").url, "mailto:")
    end
  end

  describe "links_protocol_relative" do
    test "Handles protocol-relative URLs (//example.com) correctly" do
      result = Kreuzcrawl.scrape!()
      assert length(result.links) > 1
      assert result.links.get("").protocol_relative != ""
    end
  end

  describe "links_rel_attributes" do
    test "Preserves rel=nofollow and rel=canonical attributes" do
      result = Kreuzcrawl.scrape!()
      assert length(result.links) > 0
    end
  end

  describe "links_relative_parent" do
    test "Resolves ../ and ./ relative parent path links correctly" do
      result = Kreuzcrawl.scrape!()
      assert length(result.links) > 3
    end
  end
end
